source/images.py

```diff
@@ -0,0 +1,129 @@
+        :param scale: defines the adjusted aspect ratio for the ascii image, defaults to 0.43
+        :param resolution: defines resolution, takes in values as percentage of the highest resolution possible with
+               characters from shade_str
+               Special value: 0 (default) is interpreted as mode switch to minimalistic output with characters from
+               shade_min
+        :param shade_str: defaults to a well known pattern of 69 characters used to define the greyscale
+        :param shade_min: defaults to a minimalistic pattern of 5 characters used to define the greyscale
+        """
+        # TODO: error handling when no file is selected
+        self.file = file
+        self.image = Image.open(file).convert("L")
+        self.cols = cols
+        self.scale = scale
+        self.shade = {
+            "str": shade_str,
+            "str_len": len(shade_str),
+            "min": shade_min,
+            "min_len": len(shade_min),
+            "length": 0,
+            "resolution": 0
+        }
+
+        # ensure level is [0-100]
+        resolution = 100 if resolution > 100 else 0 if resolution < 0 else resolution
+        self.shade["resolution"] = resolution
+
+    def __repr__(self) -> str:
+        return f'<Images file={self.file} image={self.image}>'
+
+    def __str__(self) -> str:
+        return '\n'.join(self.img_to_ascii)
+
+    def g_scale(self, resolution: int = 0) -> str:
+        """
+        Helper function to determine the character set based on the resolution
+
+        :return: a string of characters which corresponds to the luminance in descending order
+        """
+        if resolution == 0:
+            self.shade["length"] = self.shade["min_len"]
+            return self.shade["min"]
+        else:
+            resolution = resolution * self.shade["str_len"] // 100
+            # get evenly spaced indices and return them as a string
+            ret_scale = ""
+            indices = np.linspace(0, self.shade["str_len"] - 1, resolution, dtype=int)
+            self.shade["length"] = len(indices)
+            for i in indices:
+                ret_scale += self.shade["str"][i]
+            return ret_scale
+
+    @staticmethod
+    def get_average(image: Image) -> float:
+        """Given PIL Image, return average value of grayscale value (luminance)"""
+        # get image as numpy array and get the shape, then returns the average
+        im = np.array(image)
+        w, h = im.shape
+        return np.average(im.reshape(w * h))
+
+    @property
+    def img_to_ascii(self) -> List[str]:
+        """
+        Converts the PIL Image into a ascii picture
+
+        :return: A row-wise list of ascii characters forming the output ascii picture
+        """
+        # get the dimensions
+        horizontal, vertical = self.image.size
+        w = horizontal / self.cols
+
+        # compute tile height based on aspect ratio and scale
+        h = w / self.scale
```

Comment:
Why is one called cold but the other is called scale? How would this work with an image in another format than the one you have tested?

We should detect this value, the user enters how many columns and we figure out the width from the aspect ratio of the image opened.